File: src/research/storage/repository/write.clj
(ns research.storage.repository.write
  (:require [clojure.pprint :as pprint]
            [research.domain.session :as session]
            [research.storage.organizer :as organizer])
  (:import (java.nio.file Files StandardCopyOption)
           (java.nio.file.attribute FileAttribute)))

(defn store
  "Persist session list into output folder."
  [root items]
  (let [org (organizer/organizer root)]
    (doseq [item items]
      (let [name (organizer/name
                  org
                  (session/created item)
                  (session/topic item)
                  (session/id item))
            base (.resolve root name)
            _ (Files/createDirectories base (make-array FileAttribute 0))
            path (.resolve base "session.edn")
            text (with-out-str (pprint/pprint (session/data item)))
            temp (Files/createTempFile base "session" ".tmp"
                                       (make-array FileAttribute 0))]
        (try
          (spit (.toFile temp) text :encoding "UTF-8")
          (Files/move temp path
                      (into-array java.nio.file.CopyOption
                                  [StandardCopyOption/ATOMIC_MOVE]))
          (catch Exception exc
            (Files/deleteIfExists temp)
            (throw exc)))))))

File: crates/research-storage/src/repository/write.rs
use std::fs;
use std::path::Path;

use research_domain::session::{ResearchSession, Sessioned};

use crate::organizer::{self, Organized};

/// Persist session list into output folder.
pub fn store(root: &Path, items: &[ResearchSession]) {
    let org = organizer::organizer(root);
    for item in items {
        let name = org.name(item.created(), item.topic(), item.id());
        let base = root.join(&name);
        fs::create_dir_all(&base).ok();
        let path = base.join("session.json");
        let data = item.data();
        let text = serde_json::to_string_pretty(&serde_json::to_value(&data).unwrap_or_default())
            .unwrap_or_default();
        fs::write(&path, text).ok();
    }
}
